pkg/controller/seed-controller-manager/etcdrestore/etcd_restore_controller.go

```diff
@@ -0,0 +1,267 @@
+	}
+
+	return c.Watch(&source.Kind{Type: &kubermaticv1.EtcdRestore{}}, &handler.EnqueueRequestForObject{}, incompleteRestorePredicates)
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log := r.log.With("request", request)
+	log.Debug("Processing")
+
+	restore := &kubermaticv1.EtcdRestore{}
+	if err := r.Get(ctx, request.NamespacedName, restore); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(ctx, types.NamespacedName{Name: restore.Spec.Cluster.Name}, cluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	log = r.log.With("cluster", cluster.Name, "restore", restore.Name)
+
+	if cluster.Labels[kubermaticv1.WorkerNameLabelKey] != r.workerName {
+		return reconcile.Result{}, nil
+	}
+
+	result, err := r.reconcile(ctx, log, restore, cluster)
+	if err != nil {
+		log.Errorw("Reconciling failed", zap.Error(err))
+		r.recorder.Event(restore, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+		r.recorder.Eventf(cluster, corev1.EventTypeWarning, "ReconcilingError",
+			"failed to reconcile etcd restore %q: %v", restore.Name, err)
+	}
+
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, log *zap.SugaredLogger, restore *kubermaticv1.EtcdRestore, cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	log.Infof("performing etcd restore from backup %v", restore.Spec.BackupName)
+
+	if restore.Status.Phase == kubermaticv1.EtcdRestorePhaseCompleted {
+		return nil, nil
+	}
+
+	if restore.Status.Phase == kubermaticv1.EtcdRestorePhaseStsRebuilding {
+		return r.rebuildEtcdStatefulset(ctx, log, restore, cluster)
+	}
+
+	// pause cluster
+	if err := r.updateCluster(ctx, cluster, func(cluster *kubermaticv1.Cluster) {
+		cluster.Spec.Pause = true
+	}); err != nil {
+		return nil, fmt.Errorf("failed to pause cluster: %v", err)
+	}
+
+	if err := r.updateRestore(ctx, restore, func(restore *kubermaticv1.EtcdRestore) {
+		restore.Status.Phase = kubermaticv1.EtcdRestorePhaseStarted
+		kuberneteshelper.AddFinalizer(restore, RebuildStatefulsetFinalizer)
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add finalizer: %v", err)
+	}
+
+	// delete etcd sts
+	sts := &v1.StatefulSet{}
+	err := r.Get(ctx, types.NamespacedName{Namespace: cluster.Status.NamespaceName, Name: resources.EtcdStatefulSetName}, sts)
+	if err == nil {
+		if err := r.Delete(ctx, sts); err != nil && !kerrors.IsNotFound(err) {
+			return nil, fmt.Errorf("failed to delete etcd statefulset: %v", err)
+		}
+	} else if !kerrors.IsNotFound(err) {
+		return nil, fmt.Errorf("failed to get etcd statefulset: %v", err)
+	}
+
+	// delete PVCs
+	pvcSelector, err := labels.Parse(fmt.Sprintf("%s=%s", resources.AppLabelKey, resources.EtcdStatefulSetName))
```

Comment:
Also this was only in case of new storage class.

Reply:
see above.